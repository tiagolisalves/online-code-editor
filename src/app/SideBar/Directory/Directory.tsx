import React, { useState } from 'react'
import { FaFolderOpen, FaFolder } from 'react-icons/fa'

export function Directory(props: { name: string; children: any }) {
    const [open, setOpen] = useState(false)

    return (
        <div className="directory">
            <span onClick={() => setOpen(!open)}>
                <span className="directory-icon">{open ? <FaFolderOpen color="orange" /> : <FaFolder color="orange" />}</span>
                <span className="directory-name">{props.name}</span>
            </span>
            <DirectoryContent>{open ? props.children : ''}</DirectoryContent>
        </div>
    )
}
export function DirectoryContent(props: any) {
    return <div className="directory-content">{props.children}</div>
}
