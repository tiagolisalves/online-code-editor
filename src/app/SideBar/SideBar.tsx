import React, { useState, useContext, useEffect } from 'react'
import { IFileTree } from 'app/Models'
import { Directory } from './Directory/Directory'
import { File } from './File/File'
import { runFileTree, OpenFileContext } from './Helpers/Helpers'
import { FaTruckLoading } from 'react-icons/fa'

function SideBar(props: { fileTree: IFileTree[]; openFile: (fileId: number) => any }) {
    const [fileTree, setFileTree] = useState([] as IFileTree[])
    useEffect(() => {
        setFileTree(props.fileTree as IFileTree[])
    }, [props.fileTree])
    return (
        <OpenFileContext.Provider value={props.openFile}>
            {fileTree.length === 0 ? <span>Loading...</span> : <FileTree fileTree={fileTree}></FileTree>}
        </OpenFileContext.Provider>
    )
}

function FileTree(props: any) {
    let fileTree: IFileTree & IFileTree[] = props.fileTree
    return (
        <>
            {fileTree.length > 0 ? (
                fileTree.map((file) =>
                    (() => {
                        if (file.children?.length > 1) {
                            return (
                                <Directory key={file.id} name={file.name}>
                                    {file.children.map((child) => (
                                        <FileTree fileTree={child.isDirectory ? [child] : child} key={child.id}></FileTree>
                                    ))}
                                </Directory>
                            )
                        } else if (file.children?.length === 1) {
                            const fileTreeDirectory = runFileTree(fileTree)
                            return (
                                <Directory key={file.id} name={fileTreeDirectory[0]}>
                                    {fileTreeDirectory[1].children.map((child) => (
                                        <FileTree fileTree={child.isDirectory ? [child] : child} key={child.id}></FileTree>
                                    ))}
                                </Directory>
                            )
                        }
                    })()
                )
            ) : (
                <File key={fileTree.id} file={fileTree}></File>
            )}
        </>
    )
}

export { SideBar }
