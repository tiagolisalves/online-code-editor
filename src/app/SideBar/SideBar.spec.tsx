import { create } from 'react-test-renderer'
import React from 'react'
import { SideBar } from './SideBar'

import fixture from './fixture.json'
import { Directory } from './Directory/Directory'

test('It should get filetree from parent', () => {
    const comp = create(<SideBar fileTree={fixture as any} openFile={() => {}}></SideBar>)
    expect(comp.toJSON()).toMatchSnapshot()
})

test('It should render a directory', () => {
    let fileWasOpen = false
    const comp = create(<SideBar fileTree={fixture as any} openFile={() => (fileWasOpen = true)}></SideBar>)
    const instance = comp.root
    const directory = instance?.findAllByType(Directory)[0]
    expect(directory?.props.name).toEqual('editor / src')
})
