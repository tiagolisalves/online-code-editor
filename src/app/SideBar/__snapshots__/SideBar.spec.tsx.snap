// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should get filetree from parent 1`] = `
<span>
  Loading...
</span>
`;
